File: swift/Sources/Nestopia/Nestopia.swift
// Copyright (C) 2026 Dominic Opitz
// SPDX-License-Identifier: GPL-2.0-or-later

import SwiftUI

/// High-level SwiftUI host for one Nestopia ROM.
public struct Nestopia: View {
    @StateObject private var engine: NestopiaEngine
    private let configuration: NestopiaConfiguration
    private let controllerConfiguration: NestopiaControllerConfiguration

    public init(
        configuration: NestopiaConfiguration,
        controllerConfiguration: NestopiaControllerConfiguration = .init()
    ) {
        self.configuration = configuration
        self.controllerConfiguration = controllerConfiguration
        _engine = StateObject(wrappedValue: NestopiaEngine(configuration: configuration))
    }

    public init(
        rom: URL,
        controllerConfiguration: NestopiaControllerConfiguration = .init()
    ) {
        self.init(
            configuration: NestopiaConfiguration(romURL: rom),
            controllerConfiguration: controllerConfiguration
        )
    }

    public var body: some View {
        NestopiaView(
            engine: engine,
            showsControls: configuration.showsTouchControls,
            controllerConfiguration: controllerConfiguration
        )
        .onAppear {
            if configuration.automaticallyStarts { engine.start() }
        }
        .onDisappear { engine.stop() }
    }
}

File: android/nestopia/src/main/kotlin/net/sourceforge/nestopia/Nestopia.kt
// Copyright (C) 2026 Dominic Opitz
// SPDX-License-Identifier: GPL-2.0-or-later

package net.sourceforge.nestopia

import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.compose.LocalLifecycleOwner

@Composable
fun Nestopia(
    configuration: NestopiaConfiguration,
    modifier: Modifier = Modifier,
    controllerConfiguration: NestopiaControllerConfiguration = NestopiaControllerConfiguration(),
) {
    val context = LocalContext.current
    val engine = remember(configuration) { NestopiaEngine(context, configuration) }
    val lifecycleOwner = LocalLifecycleOwner.current

    DisposableEffect(engine, lifecycleOwner) {
        val observer =
            LifecycleEventObserver { _, event ->
                when (event) {
                    Lifecycle.Event.ON_RESUME -> engine.resume()
                    Lifecycle.Event.ON_PAUSE -> engine.pause()
                    else -> Unit
                }
            }
        lifecycleOwner.lifecycle.addObserver(observer)
        if (configuration.automaticallyStarts) engine.start()
        onDispose {
            lifecycleOwner.lifecycle.removeObserver(observer)
            engine.close()
        }
    }

    NestopiaView(
        engine = engine,
        modifier = modifier,
        showsControls = configuration.showsTouchControls,
        controllerConfiguration = controllerConfiguration,
    )
}
